### PR TITLE
Core/Misc: added new files for future global predicates and implement…

### PR DESCRIPTION
**Pull request by Ovahlord**
**Changes proposed:**

-  add a new file with a initial Predicate for future implementations of useful game predicates. Can be used later as well to clean up GridNotifiers from predicates like UnitAuraCheck

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Tests performed:** (Does it build, tested in-game, etc.)
Tested and used ingame with this scripts:

```
class spell_atramedes_sonic_breath : public SpellScript
{
    PrepareSpellScript(spell_atramedes_sonic_breath);

    void FilterTargets(std::list<WorldObject*>& targets)
    {
        if (targets.empty())
            return;

        targets.remove_if(Trinity::Predicates::UnitVictimCheck(GetCaster(), false));

        if (targets.size() > 1)
            Trinity::Containers::RandomResize(targets, 1);
    }

    void Register() override
    {
        OnObjectAreaTargetSelect += SpellObjectAreaTargetSelectFn(spell_atramedes_sonic_breath::FilterTargets, EFFECT_0, TARGET_UNIT_SRC_AREA_ENEMY);
    }
};
```

### CODE REVIEW

**Comment by jackpoz** [comment]
Circle CI reported some build warnings

File: src/server/game/Miscellaneous/CommonPredicates.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace Trinity
+{
+    namespace Predicates
+    {
+        /// Checks the given container for the attacker's current victim and either keeps it as only element or removes it
+        class TC_GAME_API UnitVictimCheck
+        {
+            public:
+                UnitVictimCheck(Unit* attacker, bool keepVictim) : _attacker(attacker), _keepVictim(keepVictim) { }
```
Comment (Treeston) [inline src/server/game/Miscellaneous/CommonPredicates.h]: heavy dislike on `bool keepVictim` to invert the operator

```cpp
template <typename PRED>
class Inverter
{
  public:
    Inverter(PRED&& p) : _child(std::move(p)) {}

    template <typename... Args>
    bool operator()(Args&&... args)
    {
      return !_child(std::forward<Args>(args)...);
    }

    template <typename... Args>
    bool operator()(Args&&... args) const
    {
      return !_child(std::forward<Args>(args)...);
    }

  private:
    PRED _child;
};

template <typename PRED>
Inverter<PRED> Invert(PRED&& p) { return Inverter<PRED>(std::forward<PRED>(p)); }
```

**Comment by ccrs** [comment]
Finally someone willing to clean this up. :+1: :+1: :+1:

**Comment by Shauren** [comment]
Why are GridNotifiers suddenly not good enough?

**Comment by Ovahlord** [comment]
Why are grid notifiers even involved for shit like UnitAuraCheck that defintively does not belong there?

**Comment by Shauren** [comment]
GridNotifiers holds two types of classes, checks/filters/predicates (pick whatever name) and visitors (you cant use a visitor without a check)
`UnitAuraCheck` was originally a grid visitor filter, uses in spell scripts came later

**Comment by Treeston** [comment]
imo unifying all generic predicates away from gridnotifiers is a good thing

**Pull request merged**